File: elixir/triangle/triangle.exs
defmodule Triangle do

  import Enum, only: [ sort: 1, dedup: 1 ]

  @type kind :: :equilateral | :isosceles | :scalene

  @doc """
  Return the kind of triangle of a triangle with 'a', 'b' and 'c' as lengths.
  """
  @spec kind(number, number, number) :: { :ok, kind } | { :error, String.t }
  def kind(a, b, c) when a <= 0 or b <= 0 or c <= 0, do: { :error, "all side lengths must be positive" }
  def kind(a, b, c) when a+b <= c or b+c <= a or a+c <= b, do: { :error, "side lengths violate triangle inequality" }
  def kind(a, a, a), do: { :ok, :equilateral }
  def kind(a, b, c), do: { :ok, [ a, b, c ] |> sort |> dedup |> check }

  defp check(sides) when length(sides) == 2, do: :isosceles
  defp check(_), do: :scalene

end

File: go/triangle/triangle.go
// Package triangle provides triabgle info functions.
package triangle

import "math"

// Kind of triangles
type Kind int

// Values for the Kind constant
const (
	NaT Kind = iota // not a triangle
	Equ             // equilateral
	Iso             // isosceles
	Sca             // scalene
)

// KindFromSides returns the kind of a triangle.
func KindFromSides(a, b, c float64) Kind {
	switch {
	case a <= 0 || b <= 0 || c <= 0: // negative sides
		return NaT
	case a+b < c || b+c < a || a+c < b: // side lengths violate triangle inequality
		return NaT
	case math.IsNaN(a + b + c): // Not a Number
		return NaT
	case math.IsInf(a+b+c, 0): // Infinite
		return NaT
	case a == b && b == c: // all sides same length
		return Equ
	case a == b || b == c || a == c: // at least two sides of same length
		return Iso
	}
	return Sca
}
